### DISCUSSION TITLE
Tokio in a web worker

### DISCUSSION THREAD

**Discussion by ardouglas** [tokio-rs/tokio#7048] [Q&A]
Hi, I've been trying to port some code I've been running as a native binary to the browser and I think I've gotten stuck.  I'm trying to initialize the runtime like: 
```
let rt = tokio::runtime::Builder::new_current_thread()
        .build_local(&LocalOptions::default())
        .unwrap();
```

but I keep getting `time not implemented on this platform` when the worker launches.  As far as I can tell from a combination of cargo hackerman explain and cargo build --timings output I have not enabled the time feature so I'm not sure where I'm going wrong.  Any help would be greatly appreciated!

**Comment by mox692** [comment]
It seems that your error is the one from standard library: 
https://github.com/rust-lang/rust/blob/d3e71fd2d3cf68addac558a807c9ba83f6217a64/library/std/src/sys/pal/unsupported/time.rs#L13

I suspect either your code or tokio is using unsupported std time feature in wasm even when tokio's time features is not used.

**Comment by Noah-Kennedy** [comment]
What are you trying to do in that code? We don't have functional timers on that platform due to platform limitations, so some things won't work right now unfortunately.

**Reply by ardouglas** [reply]
Certainly: 

```
tokio v1.42.0
├── pin-project-lite feature "default"
│   └── pin-project-lite v0.2.15
└── tokio-macros feature "default"
    └── tokio-macros v2.4.0 (proc-macro)
        ├── proc-macro2 feature "default"
        │   ├── proc-macro2 v1.0.92
        │   │   └── unicode-ident feature "default"
        │   │       └── unicode-ident v1.0.14
        │   └── proc-macro2 feature "proc-macro"
        │       └── proc-macro2 v1.0.92 (*)
        ├── quote feature "default"
        │   ├── quote v1.0.37
        │   │   └── proc-macro2 v1.0.92 (*)
        │   └── quote feature "proc-macro"
        │       ├── quote v1.0.37 (*)
        │       └── proc-macro2 feature "proc-macro" (*)
        ├── syn feature "default"
        │   ├── syn v2.0.91
        │   │   ├── proc-macro2 v1.0.92 (*)
        │   │   ├── quote v1.0.37 (*)
        │   │   └── unicode-ident feature "default" (*)
        │   ├── syn feature "clone-impls"
        │   │   └── syn v2.0.91 (*)
        │   ├── syn feature "derive"
        │   │   └── syn v2.0.91 (*)
        │   ├── syn feature "parsing"
        │   │   └── syn v2.0.91 (*)
        │   ├── syn feature "printing"
        │   │   └── syn v2.0.91 (*)
        │   └── syn feature "proc-macro"
        │       ├── syn v2.0.91 (*)
        │       ├── proc-macro2 feature "proc-macro" (*)
        │       └── quote feature "proc-macro" (*)
        └── syn feature "full"
            └── syn v2.0.91 (*)
```
I don't appear to actually be reaching any code after the attempt to start the runtime.  Were I to be able to successfully launch it though I would be running lots of small tasks that are linked together with a combination of oneshot and mpsc channels.

**Comment by ardouglas** [comment]
Update: I can see now that the panic is coming from inside of `tokio::runtime::metrics::batch::MetricsBatch::new`.
